Improve performance of NewThreadWorker, disable search for setRemoveOnCancelPolicy() on Android API < 21

artem-zinnatullin opened the pull request:
Resolves #3119.

This PR adds methods for detecting Android API Version and disables `NewThreadWorker.tryEnableCancelPolicy()` on Android API < 21 which has significant performance cost on Android.

PR also improves performance of `NewThreadWorker.tryEnableCancelPolicy()` via caching Reflection.

JakeWharton commented on src/main/java/rx/internal/schedulers/NewThreadWorker.java:
```diff
@@ -50,6 +53,16 @@
         PURGE_FREQUENCY = Integer.getInteger(FREQUENCY_KEY, 1000);
+
+        // Cached to prevent method call
+        IS_ANDROID = PlatformDependent.isAndroid();
```
This is redundant with the other checks

artem-zinnatullin replied:
I don't think so, we trying to enable cancel policy if current platform is not Android or if it's Android and API >= 21, that's why I have 2 separate booleans.

Or maybe I just don't see easier way.

JakeWharton replied:
You are checking if the API is `null` which is equivalent (and thus redundant) with this boolean.

artem-zinnatullin replied:
Yep, but then I use `IS_ANDROID` on line 127. I've improved `IS_ANDROID` initialization.

JakeWharton commented on src/main/java/rx/internal/util/PlatformDependent.java:
```diff
@@ -20,13 +20,19 @@
 
+    // Will be null if current platform is not Android
+    // Or if Android API Version can not be resolved.
+    private static final Integer ANDROID_API_VERSION = IS_ANDROID
```
Checking `IS_ANDROID` is redundant.

JakeWharton commented on src/main/java/rx/internal/util/PlatformDependent.java:
```diff
@@ -20,13 +20,19 @@
  */
 public final class PlatformDependent {
 
     private static final boolean IS_ANDROID = isAndroid0();
```
Change to `ANDROID_API_VERSION != null`

JakeWharton replied:
And delete `isAndroid0()`, there's no reason to have so many mechanisms for checking

artem-zinnatullin replied:
Ok, agree.

artem-zinnatullin commented on src/main/java/rx/internal/util/PlatformDependent.java:
```diff
@@ -20,13 +20,19 @@
 /**
  * Allow platform dependent logic such as checks for Android.
- * 
+ *
```
Uh, sorry for this, IDEA… I can fix it if you want.

artem-zinnatullin commented:
Checked on JVM (Java 8), Android 4.1.2 (API 16), Android 5.1.1 (API 22).

akarnokd commented on src/main/java/rx/internal/util/PlatformDependent.java:
```diff
@@ -20,31 +20,49 @@
-    private static final boolean IS_ANDROID = isAndroid0();
+    // Will be null if current platform is not Android
+    // Or if Android API Version can not be resolved.
+    private static final Integer ANDROID_API_VERSION = resolveAndroidApiVersion();
```
Is there an API version 0? If not, we could use `int` here so there is no need for unboxing and null checking elsewhere.

artem-zinnatullin replied:
Nope, there is no API version 0 as far as I know, we can also use `-1` as Not-Android-Platform value and even create a constant.

JakeWharton replied:
Nope. Great idea! RxJava doesn't support anything below API 9 anyway.

JakeWharton commented on src/main/java/rx/internal/schedulers/NewThreadWorker.java:
```diff
@@ -50,6 +53,15 @@
+        final Integer androidApiVersion = PlatformDependent.getAndroidApiVersion();
+
+        // Cached to prevent method call
+        IS_ANDROID = androidApiVersion != null;
```
You can still eliminate this variable altogether. The same check is included in the other.

JakeWharton commented on src/main/java/rx/internal/schedulers/NewThreadWorker.java:
```diff
@@ -112,7 +124,7 @@ static void purgeExecutors() {
      */
     public static boolean tryEnableCancelPolicy(ScheduledExecutorService exec) {
-        if (!PURGE_FORCE) {
+        if (!PURGE_FORCE && (!IS_ANDROID || IS_ANDROID_AND_SUPPORTS_SET_REMOVE_ON_CANCEL_POLICY)) {
```
What i'm saying is, just make this 
```java
if (!PURGE_FORCE && IS_ANDROID_AND_SUPPORTS_SET_REMOVE_ON_CANCEL_POLICY)
```
to avoid the redundancy of having both variables and checking the same condition twice on the true case.

artem-zinnatullin replied:
Maybe I'm not understanding something because it's 10 pm and I'm still in the office... 

But your code will always return `false` on regular JVM because `IS_ANDROID_AND_SUPPORTS_SET_REMOVE_ON_CANCEL_POLICY` will be `false`.

JakeWharton replied:
Should be an OR and a not then.

JakeWharton replied:
Sorry my boolean logic is all screwed up because I have my own version of this in my head.

My point was and remain, both of these are constant booleans which means whatever logic you want to apply to them can be done once. This yields a single static field which encapsulates both conditions (not Android or API 22+).

akarnokd replied:
I'd change the entire method as well to do some caching and avoid looping at all if applied to the default `ScheduledThreadPoolExecutor` that is returned from `Executors.newScheduledThreadPool()`:

```java
private static volatile Object setOnCancelPolicyMethod;
private static final Object SET_ON_CANCEL_NOT_SUPPORTED = new Object();

public static boolean tryEnableCancelPolicy(ScheduledExecutorService exec) {
    if (!PURGE_FORCE && NOT_ANDROID_OR_SUPPORTS_SET_ON_CANCEL_POLICY) {
        final boolean executorPool = exec instanceof ScheduledThreadPoolExecutor;
        Method methodToCall;
        if (executorPool) {
            final Object localSetOnCancelPolicyMethod = setOnCancelPolicyMethod;
            if (localSetOnCancelPolicyMethod == SET_ON_CANCEL_NOT_SUPPORTED) {
                return false;
            }
            if (localSetOnCancelPolicyMethod == null) {
                Method m = findSetOnCancelPolicyMethod(exec);
                
                setOnCancelPolicyMethod = m != null ? m : SET_ON_CANCEL_NOT_SUPPORTED;
                
                methodToCall = m;
            } else {
                methodToCall = (Method)localSetOnCancelPolicyMethod;
            }
        } else {
            methodToCall = findSetOnCancelPolicyMethod(exec);
        }
        
        if (methodToCall != null) {
            try {
                methodToCall.invoke(exec, true);
                return true;
            } catch (Exception ex) {
                RxJavaPlugins.getInstance().getErrorHandler().handleError(ex);
            }
        }

    }
    return false;
}

private static Method findSetOnCancelPolicyMethod(ScheduledExecutorService exec) {
    for (final Method m : exec.getClass().getMethods()) {
        if (m.getName().equals("setRemoveOnCancelPolicy")
                && m.getParameterTypes().length == 1
                && m.getParameterTypes()[0] == Boolean.TYPE) {
            return m;
        }
    }
    return null;
}
```

artem-zinnatullin replied:
@JakeWharton done! Take a look at `SHOULD_TRY_ENABLE_CANCEL_POLICY`.

@akarnokd done! Rewritten your code, added some tests and checked that it handles inheritance for `ScheduledExecutorService` (I was afraid that cached method won't work for child classes).

And tested again on JVM (Java 8), Android API 19 and Android API 22.

I should say that I noticed performance improvement in production Android app after these change for Android API < 21, hope it'll be in RxJava 1.0.14.

artem-zinnatullin replied:
Also, I used `int` for Android API Version. And as far as I saw via debugging this method called in many cases, so this PR should improve performance of RxJava in many apps (backends and clients).

JakeWharton commented on src/main/java/rx/internal/util/PlatformDependent.java:
```diff
@@ -20,31 +20,56 @@
-    private static final boolean IS_ANDROID = isAndroid0();
+    /**
+     * Constant value of Android API Version which means that current platform is not Android.
+     * We could use {@link Integer} and {@code null}, but we want to avoid boxing/unboxing.
```
Second sentence is not useful to retain.

JakeWharton commented on src/main/java/rx/internal/schedulers/NewThreadWorker.java:
```diff
@@ -48,8 +49,17 @@
         PURGE_FREQUENCY = Integer.getInteger(FREQUENCY_KEY, 1000);
+
+        // Forces the use of purge even if setRemoveOnCancelPolicy is available
+        final boolean purgeForce = Boolean.getBoolean(PURGE_FORCE_KEY); // Compiler should inline this
```
I don't understand what this trailing comment is telling me. Why is it here? Speculation about compiler optimizations make for bad comments. Especially since this is a local field in a static block that's only ever run once where perf doesn't _really_ matter.

JakeWharton commented on src/main/java/rx/internal/schedulers/NewThreadWorker.java:
```diff
@@ -48,8 +49,17 @@
+        final int androidApiVersion = PlatformDependent.getAndroidApiVersion();
+
+        SHOULD_TRY_ENABLE_CANCEL_POLICY = !purgeForce
+                        && (androidApiVersion == ANDROID_API_VERSION_IS_NOT_ANDROID || androidApiVersion >= 21);
```
Four space indent is already killing me on this whole project, but continuations should be double that which is 8. This is 16.

artem-zinnatullin replied:
+1, my bad. Would be nice to switch to 2 spaces instead of 4.

JakeWharton commented on src/main/java/rx/internal/schedulers/NewThreadWorker.java:
```diff
@@ -48,8 +49,17 @@
+        final boolean purgeForce = Boolean.getBoolean(PURGE_FORCE_KEY); // Compiler should inline this
+
+        // According to http://developer.android.com/reference/java/util/concurrent/ScheduledThreadPoolExecutor.html#setRemoveOnCancelPolicy(boolean)
+        // setRemoveOnCancelPolicy available since Android API 21
```
These comments belong on the static field initialization where the comparison against 21 is actually made.

akarnokd commented on src/test/java/rx/internal/schedulers/NewThreadWorkerTest.java:
```diff
@@ -0,0 +1,103 @@
+        boolean result = NewThreadWorker.tryEnableCancelPolicy(executor);
+
+        assertTrue(result);
+        verify(executor).setRemoveOnCancelPolicy(true);
```
This will fail on a Java 6 build but I'm not sure if people actually build RxJava from source on Java 6. What I would do is subclass ScheduledThreadPoolExecutor and add the setRemoveOnCancelPolicy(boolean) method. For Java 7+, this will count as a simple override.

artem-zinnatullin replied:
Nice workaround!

akarnokd commented on src/test/java/rx/internal/schedulers/NewThreadWorkerTest.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    @Test
+    public void tryEnableCancelPolicyShouldBeSkipped() throws NoSuchFieldException, IllegalAccessException {
```
This looks rather dangerous. What I'd do is to have a package-private method `tryEnableCancelPolicy(ScheduledExecutor executor, boolean tryEnable)` so the test can inject the necessary value.

artem-zinnatullin replied:
Hm, a separate method just for injecting value which can completely disable it… Not sure I am following this (on Android we have some limits on the quantity of methods in the app, this will increment it just for test purposes), let's just remove this test for now? 

Or I can remove `private` modifier from `SHOULD_TRY_ENABLE_CANCEL_POLICY` and make this test more "static", reflection will only change the value of this field.

akarnokd replied:
I'd rather not have this test for now. All other changes appear to be okay. If you update the PR, I'll merge it.

artem-zinnatullin commented:
Fixed all comments, removed test with reflection, added more javadoc and comments, tests are now compilable on JDK 6 (didn't check but they should work fine). @akarnokd, @JakeWharton PTAL

akarnokd commented on src/main/java/rx/internal/util/PlatformDependent.java:
```diff
@@ -20,31 +20,55 @@
+    /**
+     * Resolves version of Android API.
+     *
+     * @return version of Android API or {@code null} if version can not be resolved.
```
This no longer returns null but `ANDROID_API_VERSION_IS_NOT_ANDROID`.

artem-zinnatullin replied:
Fixed.

akarnokd commented:
Great, thanks!
